reporting/performance/prime_sieve_performance.cpp

```diff
@@ -0,0 +1,133 @@
+
+#include <boost/math/special_functions/prime_sieve.hpp>
+#include <boost/math/special_functions/interval_sieve.hpp>
+#include <benchmark/benchmark.h>
+#include <primesieve.hpp>
+#include <vector>
+
+// Individual Algos
+template<class Integer>
+inline auto linear_sieve_helper(Integer upper_bound, std::vector<Integer> primes) -> std::vector<Integer>
+{
+    boost::math::detail::linear_sieve(upper_bound, primes);
+    return primes;
+}
+
+template<class Integer>
+void linear_sieve(benchmark::State& state)
+{
+    Integer upper = static_cast<Integer>(state.range(0));
+    for(auto _ : state)
+    {
+        std::vector<Integer> primes;
+        benchmark::DoNotOptimize(linear_sieve_helper(upper, primes));
+    }
+    state.SetComplexityN(state.range(0));
+}
+
+template<class Integer>
+inline auto mask_sieve_helper(Integer lower_bound, Integer upper_bound, std::vector<Integer> primes) -> std::vector<Integer>
+{
+    boost::math::detail::mask_sieve(lower_bound, upper_bound, primes);
+    return primes;
+}
+
+template<class Integer>
+void mask_sieve(benchmark::State& state)
+{
+    Integer lower = static_cast<Integer>(2);
+    Integer upper = static_cast<Integer>(state.range(0));
+    for(auto _ : state)
+    {
+        std::vector<Integer> primes;
+        benchmark::DoNotOptimize(mask_sieve_helper(lower, upper, primes));
+    }
+    state.SetComplexityN(state.range(0));
+}
+
+template<class Integer>
+inline auto interval_sieve_helper(Integer lower_bound, Integer upper_bound, std::vector<Integer> primes) -> std::vector<Integer>
+{
+    std::vector<Integer> pre_sieved_primes {2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71};
+    boost::math::detail::IntervalSieve sieve(lower_bound, upper_bound, pre_sieved_primes, primes);
+    return primes;
+}
+
+template<class Integer>
+void interval_sieve(benchmark::State& state)
+{
+    Integer lower = static_cast<Integer>(2);
+    Integer upper = static_cast<Integer>(state.range(0));
+    for(auto _ : state)
+    {
+        std::vector<Integer> primes;
+        benchmark::DoNotOptimize(interval_sieve_helper(lower, upper, primes));
+    }
+    state.SetComplexityN(state.range(0));
+}
+
+// Complete Implementations
+template<class ExecuitionPolicy, class Integer, class Container>
+inline auto prime_sieve_helper(ExecuitionPolicy policy, Integer upper, Container primes)
+{
+    boost::math::prime_sieve(policy, upper, primes);
+    return primes;
+}
+
+template <class Integer>
+void prime_sieve(benchmark::State& state)
+{
+    Integer upper = static_cast<Integer>(state.range(0));
+    for(auto _ : state)
+    {
+        std::vector<Integer> primes;
+        benchmark::DoNotOptimize(prime_sieve_helper(std::execution::par, upper, primes));
+    }
+    state.SetComplexityN(state.range(0));
+}
+
+template <class Integer>
+inline auto kimwalish_primes_helper(Integer upper, std::vector<Integer> primes) -> std::vector<Integer>
+{
+    primesieve::generate_primes(upper, &primes);
```

Comment:
Where is this `generate_primes` function? Is it not part of this PR?

Reply:
This is part of a separate libray which is used to compare performance: https://github.com/kimwalisch/primesieve

Reply:
Aha. 

So is the BSD 2-Clause license not compatible with the Boost license?

Reply:
@jeremy-murphy : This is only *using* the sieve in the performance section; there should be no licensing problems.

Reply:
Ah, what I meant is, can we not take KW's code directly because of the license? Is that why we're writing our own?

Reply:
I was not aware of kimwalisch when this started. I would say the major distinguishing factor now is that he is limited to 2^64 as an upper bound where this supports arbitrary precision through the boost multi-precision back-ends.